core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -52,6 +52,7 @@ import scala.math._
 abstract class AbstractFetcherThread(name: String,
                                      clientId: String,
                                      val sourceBroker: BrokerEndPoint,
+                                     markPartitionsFailed: Set[TopicPartition] => Unit,
```

Comment:
Was there any reason to accept a `Set` rather than a single partition?

Reply:
By the way, I wonder if we may as well put a `trait` behind this. That would be a little test-friendly.

```
trait FailedPartitions {
  def size: Int
  def add(partition: TopicPartition): Unit
  def remove(partition: TopicPartition): Unit
}
```
We could protect the implementation with a separate lock, which might solve the locking issue mentioned elsewhere.

Reply:
Made a class for FailedPartitions with the necessary functions.